dask/layers.py

```diff
@@ -154,17 +157,38 @@ def __iter__(self):
         return iter(self._dict)
 
     def __len__(self):
-        return len(self._dict)
+        interior_keys, _ = self._keys_from_overlap_operation()
+        n_tasks = 0
+        for k in interior_keys:
+            if all([isinstance(i, int) for i in k]):
+                n_tasks += 2  # represents both getitem and overlap tasks
+            else:
+                n_tasks += 1  # represents a single getitem task
+        return n_tasks
 
     def is_materialized(self):
         return hasattr(self, "_cached_dict")
 
     def get_output_keys(self):
-        return self.keys()  # FIXME! this implementation materializes the graph
+        if self._cached_output_keys is not None:
+            return self._cached_output_keys
 
-    def _dask_keys(self):
-        if self._cached_keys is not None:
-            return self._cached_keys
+        getitem_name = "getitem-" + self.token
+        overlap_name = "overlap-" + self.token
+        interior_keys, _ = self._keys_from_overlap_operation()
+        interior_slices = {}
+        overlap_blocks = {}
+        for k in interior_keys:
+            interior_slices[(getitem_name,) + k] = None
+            if all([isinstance(i, int) for i in k]):
+                overlap_blocks[(overlap_name,) + k] = None
+        output_keys = toolz.merge(interior_slices, overlap_blocks).keys()
+        self._cached_output_keys = output_keys  # cache output_keys
+        return output_keys
+
+    def _keys_from_input_dependencies(self):
```

Comment:
This feels like it could be called `_get_input_keys` to make the relationship with `get_output_keys` clearer